docs/dg/dev/architecture/dependency-injection/best-practices.md

```diff
@@ -0,0 +1,453 @@
+
+
+### Keep services focused (Single Responsibility)
+
+Each service should have one clear purpose:
+
+```php
+<?php
+
+// ✅ Good: Focused services
+class CustomerEmailValidator
+{
+    public function validate(string $email): bool
+    {
+        // Only validates emails
+    }
+}
+
+class CustomerAddressValidator
+{
+    public function validate(AddressTransfer $address): bool
+    {
+        // Only validates addresses
+    }
+}
+
+// ❌ Bad: Too many responsibilities
+class CustomerValidator
+{
+    public function validateEmail(string $email): bool { }
+    public function validateAddress(AddressTransfer $address): bool { }
+    public function validatePhone(string $phone): bool { }
+    public function validateCreditCard(string $card): bool { }
+    // Too many unrelated validations!
+}
+```
+
+## Performance considerations
+
+### Use service proxies for expensive services
+
+For services that are expensive to instantiate but not always needed:
+
+```php
+<?php
+
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+
+return static function (ContainerConfigurator $configurator): void {
+    $services = $configurator->services()
+        ->defaults()
+        ->autowire()
+        ->public()
+        ->autoconfigure();
+
+    // Mark service as lazy to create proxy
+    $services->set(ExpensiveService::class)
+        ->lazy();
+};
+```
+
+### Avoid service explosion
+
+Don't create a service for every single class. Use factories for lightweight objects:
+
+```php
+<?php
+
+// ✅ Good: Use factory for simple value objects
+class CustomerFactory
+{
+    public function createCustomerData(array $data): CustomerData
+    {
+        return new CustomerData(
+            $data['name'],
+            $data['email']
+        );
+    }
+}
+
+// ❌ Bad: Making every value object a service
+$services->set(CustomerData::class); // Unnecessary!
+```
+
+### Consider container compilation time
+
+Each service registered increases compilation time. For development:
+
+- Exclude heavy modules (DataImport, Storage, Search)
+- Use container cache
+- Compile container in deployment pipeline
+
+```php
+<?php
+
+$excludedModuleConfiguration = [
+    'DataImport' => true,
+    'ProductPageSearch' => true,
+    'ProductStorage' => true,
+    'PriceProductStorage' => true,
+];
+```
+
+## Testing best practices
+
+### Use constructor injection for better testability
+
+```php
+<?php
+
+namespace Pyz\Zed\Customer\Business;
+
+class CustomerService
+{
+    public function __construct(
+        private CustomerRepositoryInterface $repository,
+        private EmailValidatorInterface $validator
+    ) {
+    }
+
+    public function registerCustomer(CustomerTransfer $customer): bool
+    {
+        if (!$this->validator->validate($customer->getEmail())) {
+            return false;
+        }
+
+        $this->repository->save($customer);
+
+        return true;
+    }
+}
+```
+
+## Migration checklist
```

Comment:
How is migration related to the best practices?

I suggest to move this to a new page